compendium/ConfigurationAdmin/src/CMAsyncWorkService.cpp

```diff
@@ -27,167 +27,169 @@
-    // replace existing asyncWorkService with a nullptr asyncWorkService
-    std::shared_ptr<cppmicroservices::async::AsyncWorkService> newService =
-      std::make_shared<FallbackAsyncWorkService>(logger);
-    std::atomic_store(&asyncWorkService, newService);
-  }
-}
-
-void CMAsyncWorkService::post(std::packaged_task<void()>&& task)
-{
-  auto currAsync = std::atomic_load(&asyncWorkService);
-  currAsync->post(std::move(task));
-}
-
-}
-}
+    namespace cmimpl
+    {
+
+        /**
+         * FallbackAsyncWorkService represents the fallback strategy in the event
+         * that a AsyncWorkService is not present within the framework. It implements
+         * the public interface for AsyncWorkService and is created in the event that
+         * a user-provided service was not given or if the user-provided service
+         * which implements the AsyncWorkService interface was unregistered.
+         */
+        class FallbackAsyncWorkService final : public cppmicroservices::async::AsyncWorkService
+        {
+          public:
+            FallbackAsyncWorkService(std::shared_ptr<cppmicroservices::logservice::LogService> const& logger_)
+                : logger(logger_)
+            {
+                Initialize();
+            }
+
+            void
+            Initialize()
+            {
+                threadpool = std::make_shared<boost::asio::thread_pool>(1);
+            }
+
+            void
+            Shutdown()
+            {
+                if (threadpool)
+                {
+                    try
+                    {
+                        threadpool->join();
+                        threadpool->stop();
+                        threadpool.reset();
+                    }
+                    catch (...)
+                    {
+                        auto exceptionPtr = std::current_exception();
+                        std::string msg = "An exception has occurred while trying to shutdown "
+                                          "the fallback cppmicroservices::async::AsyncWorkService "
+                                          "instance.";
+                        logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_WARNING, msg, exceptionPtr);
+                    }
+                }
+            }
+
+            ~FallbackAsyncWorkService() { Shutdown(); }
+
+            void
+            post(std::packaged_task<void()>&& task) override
+            {
+                if (threadpool)
+                {
+                    using Sig = void();
+                    using Result = boost::asio::async_result<decltype(task), Sig>;
+                    using Handler = typename Result::completion_handler_type;
+
+                    Handler handler(std::forward<decltype(task)>(task));
+                    Result result(handler);
+
+                    boost::asio::post(threadpool->get_executor(),
+                                      [handler = std::move(handler)]() mutable { handler(); });
```

Comment:
What's going on here? "handler = std::move(handler)"???

Reply:
@carneyweb please create a github issue for this

Reply:
From what I remember when looking at this code, the `handler = std::move(handler)` was a direct port from what we were originally using w.r.t. the ASIO threadpool in DS. See the previous code here: https://github.com/CppMicroServices/CppMicroServices/commit/bd4478c10cf0a9da62686b8879528a3850ca3b60#diff-3843ccf1a1f1641bbd12616e66a9f4b2a1203a5d888cca3c9ade6a15f4592251L179 .

Not sure if this is an "issue" per-se... as I know we had to do some _wacky_ things to get around a MSVC bug regarding lambda captures and it not properly implicitly capturing a moved type.

There is a more "interesting" example of this here (https://github.com/CppMicroServices/CppMicroServices/blob/development/compendium/DeclarativeServices/src/manager/ComponentManagerImpl.cpp#L165).

Reply:
This `handler = std::move(handler)` will exist in all of the `*AsyncWorkService.cpp` files as they all are copies of each other with the names changed.

I think at some point we should explore having a unified implementation for DS, CA, other compendium services, and (maybe) the core framework. As of right now, when DS and CA both start, we actually create 2 unique thread pools rather than 1 shared one.